Shooter pull req

lilbites21 opened the pull request:
i think i have all the necessary things for my subsystem. we would need to fine tune the const. i kind of just guessed what would be good. and maybe change the name of some variables if they contradict, or just suck(I am not good with var names).

lilbites21 committed:
adding shooter subsystem

added ShooterMap
added mock map for Gaston

lilbites21 committed:
change velocity of ball

changes the shooter motor's speed based on how far down LTrigger is pushed. finer control at smaller speeds

lilbites21 committed:
added PID controll, i think;

checks if if actual speed is close to target speed, add added a command that changes speed based on the difference of target and actual speed

lilbites21 committed:
shoot controls

PID control can actually change motor speed now, and when b is pressed, will make sure that the speed is what it is suppose to be, them shoots(shooting is a work in progress)

lilbites21 committed:
added Shoot command

changed shoot button from b to rBumper, and can now shoot ball by turning on intakeMotor for a few seconds then turning it off

lilbites21 committed:
changing motor speed easier

cleaned up the code by moving the commands and methods to the end, and now change motor speed periodically using periodic method instead of changing it every frame whenever i wanted to change it.

lilbites21 committed:
fixxed checkspeed

realized that the last implimentation wouldnt do anything, so i corrected it so it would do the proper function. i think

lilbites21 committed:
changed speed and power to 2 diffrent things

i mad it so speed is the var that is the speed you want and power is the number you give to the motor. lets you know what the speed you want is wile you change the actual speed.

lilbites21 committed:
dont know why this was in this branch

no need for this in this branch

lilbites21 committed:
Change wait time based on error.

The closer we are to the target speed. the less we have to wait.

lilbites21 committed:
Formatted code, and changed pilot speed control.

Changed how pilot changes speed by having to press left bumper. It's so that the the speed doesn't change without the pilots consent.

lilbites21 committed:
Merge branch 'main' into Shooter

lilbites21 committed:
Got rid of not used motors. fixed getShooterMap

Got rid of lateral and longitudinal motors. and changed getShooterMap to what everyone else was doing because that was the correct way of doing it

lilbites21 committed:
Formatted code, added test motors, and added speed to dashboard

summary is self explanitory

lilbites21 committed:
Merge remote-tracking branch 'origin/main' into Shooter

lilbites21 committed:
Added safeState

turns off motor in safeState method

lilbites21 committed:
Added set default speed command and formating

Formatted Robot.java to make it more readable. binded upper goal speed and lower goal speed to UP and DOWN buttons

lilbites21 committed:
Got rid of "annoying" symbols in my comments

lilbites21 committed:
actually got rid of symbols

oops

lilbites21 committed:
Changed shoot command

used built in command base method method rather than making a new command base object

lilbites21 committed:
added proper pid controll

used built in pid controller in the motor instead of using my own rag tag method.

lilbites21 committed:
no null pointer exeption

assign shoot encoder in constructor after shooter motor gets assigned

msoucy commented on simgui-ds.json:
```diff
@@ -0,0 +1,98 @@
+{
```
Do we want these files saved? Or are they specific for a test?

msoucy commented on simgui.json:
```diff
@@ -0,0 +1,48 @@
+{
```
Do we want these values saved? Or are they specific for a test?

msoucy commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -23,6 +28,8 @@
 
   private final Intake intake = new Intake(map.getIntakeMap());
 
+  private Shooter shooter = new Shooter(map.getShooterMap()); // gets shootermap and makes a shooter subsystem
```
Comment isn't really necessary here - comments should generally explain why you're doing something this way

msoucy commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -91,6 +92,16 @@ public SwerveDriveMap getSwerveDriveMap() {
+
+    public ShooterMap getShooterMap() {
+        shooterMoter.setControlType(PIDControlType.Velocity);
+        ShooterMap map = new ShooterMap(shooterMoter, shooterIntakeMoter);
```
Unnecessary temporary variable.

msoucy commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -91,6 +92,16 @@ public SwerveDriveMap getSwerveDriveMap() {
     private final SparkMaxPIDController deploymentPidController = deploymentMotor.getPidController();
 
+    // shooter stuff
+    private final PIDSparkMax shooterMoter = new PIDSparkMax(5, MotorType.kBrushless);
```
Please construct these inside `getShooterMap`

msoucy commented on vendordeps/PhotonLib-json-1.0.json:
```diff
@@ -0,0 +1,41 @@
+{
```
Is this necessary as part of this pull request?

msoucy replied:
@lilbites21 ???

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+    double velocity = shootWheelWidth * Math.PI * shootEncoder.getRate() / (60 * 12);
+    // diameter*pi(curcumfrence)* rmp (inches per min) / 60 (inches per second)
+    // / 12 (feet per second)
+    shooterMotor.setSetpoint(shootSpeed);
```
Do not set the setpoint in periodic - it'll overwrite the speeds that you set below.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  public CommandBase shoot() {
+    return sequence("Shoot", setIntakeSpeed(INTAKESPEED), new WaitCommand(SHOOTTIME), setIntakeSpeed(0.0));
```
Why does `shoot` mess with the intake motor, not the shooter motor?

lilbites21 replied:
the shoot motor should be up to speed at the point of shooting. it turns on the intake motor to transport the ball from the push the ball into the shoot motor. I think that is all it needs to do. if I am wrong ill change it, but I don't wee why the shoot needs to change shoot motor.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+    return this.waitTime;
+  }
+
+  public CommandBase setDefaultSpeed(int index) {
```
Please use an enum for this. Do not use a hidden array and index into it, it doesn't provide clear names.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+
+  public CommandBase setSpeed(DoubleSupplier speed) {
+    return running("setSpeed", () -> {
+      setSpeedF(speed.getAsDouble() * speed.getAsDouble());
```
Calling `getAsDouble()` twice could actually give two results. Recommend calling it once and using `Math.pow` or storing it in a variable and multiplying that by itself.

msoucy commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -33,12 +40,40 @@ public void robotInit() {
+    copilotController.lbumper().whileHeld(shooter.setSpeed(copilotController::getLeftTriggerAxis));
+    // when left bumper is pressed lets pilot set speed :)
+    copilotController.rbumper().whenPressed(sequence("Shoot",
+        new WaitCommand(shooter.getWaitTime()), // gives the PID controll time to do its thing
```
Could you write a command that waits until the PID controller is at the setpoint?

bot190 replied:
See https://github.com/chopshop-166/frc-2020/blob/main/src/main/java/frc/robot/subsystems/Shooter.java#L113 for how we've done this in the past.

bot190 commented on src/main/java/frc/robot/maps/RobotMap.java:
```diff
@@ -136,6 +136,28 @@ public BooleanSupplier getOutsideLimit() {
 
+    public class ShooterMap {
+        private SmartMotorController shootMotor;
+        private SmartMotorController intakeMotor;
```
The shooter map shouldn't have intake related functionality. This is taken care of by the intake subsystem.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+  private double defaultSpeedNums[] = { 0.0, 0.5, 1.0 };
+  private double shootSpeed;
+
+  private double shootWheelWidth = 2;
```
Isn't this diameter not width? And this type of data should be in the robot specific map (`GastonMap.java`).

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,88 @@
+
+  @Override
+  public void periodic() {
+    double velocity = shootWheelWidth * Math.PI * shootEncoder.getRate() / (60 * 12);
```
This conversion should be done in robot map so that `shootEncoder` returns the correct units (Why not stick to RPM?)

msoucy replied:
`motor.getEncoder().setPositionScaleFactor(magic_math)`

lilbites21 committed:
Deleted unnecessary files

Accidentally added 2 files. Got rid of them!

lilbites21 committed:
Changed the robot and Gaston map

put variables in the correct spot, added wheel diameter to a robot specific map. changed "inputMotor", and everything else named input, to loading to stop confusion.

lilbites21 committed:
shooting now waits for shoot motor to speed up.

when shooting, it now waits for the motor to speed up to the speed we want. also changed "inputMotor", and everything else named input, to loading to stop confusion. made default speeds a enum to be more readable!!! :)

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,110 @@
+
+  private double shootWheelWidth;
+  private double velMul;
+  private double MAXRPM = 5000;
```
Some of these are constants and should be marked (and named) as such.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  public CommandBase waitTilSpeedUp() {
+    BooleanSupplier somethingMoreCohearant = () -> Math.abs(shootEncoder.getRate())
```
What is this variable name, other than misspelled?

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,110 @@
+  public CommandBase setSpeed(DoubleSupplier speed) {
+    double speeds = speed.getAsDouble() * MAXRPM;
+    return running("setSpeed", () -> {
+      shooterMotor.setSetpoint(speeds * speeds);
```
You're potentially squaring the max RPM? And only calling the speed `function` once?

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,110 @@
+  @Override
+  public void periodic() {
+    double velocity = shootEncoder.getRate() * velMul;
+    // diameter*pi(curcumfrence)* rmp (inches per min) / 60 (inches per second)
```
Stray comment?

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,110 @@
+    loaderMotor = shooterMap.getLoadingMotor();
+    shootWheelWidth = shooterMap.getWheelDiameter();
+    shootEncoder = shooterMotor.getEncoder();
+    System.out.print(DefaultSpeed.OFF);
```
Don't do `System.out.print()` in robot code

msoucy commented on src/main/java/frc/robot/maps/RobotMap.java:
```diff
@@ -136,6 +136,35 @@ public BooleanSupplier getOutsideLimit() {
+        private double shootWheelDiameter;
+
+        public ShooterMap(SmartMotorController shootMotorController, SmartMotorController loadingMotorController,
+                double shootWheelSize) {
```
What unit is `shootWheelSize` in?

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  public double getWaitTime() {
+    return this.waitTime;
```
Is this used anywhere?

lilbites21 committed:
Named constants correctly, and formating.

lilbites21 committed:
Added more self-explanatory names

lilbites21 committed:
Spelled check correctly

msoucy commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -91,6 +92,14 @@ public SwerveDriveMap getSwerveDriveMap() {
     private final SparkMaxPIDController deploymentPidController = deploymentMotor.getPidController();
 
+    public ShooterMap getShooterMap() {
+        final PIDSparkMax shooterMoter = new PIDSparkMax(5, MotorType.kBrushless);
```
These ID should not conflict with the drive IDs.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,107 @@
+  public CommandBase setDefaultSpeed(DefaultSpeed speed) {
+    return instant("Set default speed", () -> {
+      shooterMotor.setSetpoint(speed.getSpeed() * MAX_RPM);
+      shootSpeed = speed.getSpeed() * MAX_RPM;
```
Why are you setting `shootSpeed` here?

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,107 @@
+
+  public CommandBase waitTilSpeedUp() {
+    BooleanSupplier check = () -> Math.abs(shootEncoder.getRate())
+        - Math.min(MAX_RPM, shootSpeed) < RPM_BUFFER;
```
`shootSpeed` never changes here - is that a problem?

lilbites21 replied:
it doesn't need to change the speed, the PID has that under control. it just waits for it to hit the target speed.

msoucy commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -33,12 +39,39 @@ public void robotInit() {
+    // sets lower goal speed
+    copilotController.lbumper().whileHeld(shooter.setSpeed(copilotController::getLeftTriggerAxis));
+    // when left bumper is pressed lets pilot set speed :)
+    copilotController.rbumper().whenPressed(sequence("Shoot",
```
If this whole sequence is shooter-specific, can it be created in the shooter subsystem?

jeffaburke commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -91,6 +92,14 @@ public SwerveDriveMap getSwerveDriveMap() {
+    public ShooterMap getShooterMap() {
+        final PIDSparkMax shooterMoter = new PIDSparkMax(5, MotorType.kBrushless);
+        final PIDSparkMax shooterLoadingMoter = new PIDSparkMax(6, MotorType.kBrushless);
+        final double shootWheelDiameter = 2.0;
```
What units is this in?

JoeJanickiChSh committed:
Merge branch 'main' into Shooter

JoeJanickiChSh committed:
Fix naming

JoeJanickiChSh committed:
Motor IDs

JoeJanickiChSh committed:
Naming and formatting changes

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  public CommandBase setSpeed(DoubleSupplier speed) {
+    double speeds = speed.getAsDouble();
```
Is this supposed to be getting the speed only once, or on each loop? If once, why is it a supplier?

lilbites21 replied:
The speed its based on the the xbox trigger. so it gets called every frame. I did change the command type so maybe it was bad before that.

bot190 replied:
This function is only called once when the robot starts so you can't get the speed here. If this was in the instant command itself it would get a new speed everytime the command runs.

JoeJanickiChSh committed:
Multiply radius by 2

That is diameter

JoeJanickiChSh committed:
Move velocity multiplier into RobotMap

lilbites21 committed:
Moved the complexity of shooting to ShootMap.

lilbites21 committed:
changed setSpeed to an instant command

it sets the speed once. I don't know why it was a running command

lilbites21 committed:
Merge remote-tracking branch 'origin/main' into Shooter

lilbites21 committed:
💾

lilbites21 committed:
Got rid of duplicate line

⠀

lilbites21 committed:
Merge remote-tracking branch 'origin/Shooter' into Shooter

msoucy commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -70,6 +70,19 @@ public SwerveDriveMap getSwerveDriveMap() {
+        final double shootWheelRadius = 2.0; // Radius in inches
+
+        // Convert to rotations per minute?
+        final double velocityMultiplier = shootWheelRadius * 2 * Math.PI / (60 * 12);
```
- Where does the 12 come from?
- I suggest using `setPositionScaleFactor()` and `setVelocityScaleFactor()` for this.

bot190 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -35,11 +38,26 @@ public void robotInit() {
     copilotController.a().whileHeld(intake.extend(SpinDirection.COUNTERCLOCKWISE));
     copilotController.b().whileHeld(intake.retract(SpinDirection.COUNTERCLOCKWISE));
 
+    // Drive:
+    driveController.back().whenPressed(drive.resetCmd());
+    driveController.back().whenPressed(drive.resetCmd());
```
Why are there drive changes in the shooter pr?

msoucy commented on src/main/java/frc/robot/maps/RobotMap.java:
```diff
@@ -136,6 +136,41 @@ public BooleanSupplier getOutsideLimit() {
         }
     }
 
+    public class ShooterMap {
```
Please move `ShooterMap` to the `frc.robot.maps.subsystems` package.

jeffaburke replied:
What does this mean? (Putting this off until later)

bot190 commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -70,6 +70,19 @@ public SwerveDriveMap getSwerveDriveMap() {
 
+    @Override
+    public ShooterMap getShooterMap() {
+        final PIDSparkMax shooterMoter = new PIDSparkMax(13, MotorType.kBrushless);
```
There are two motors for the shooter. I think the IDs are 15 and 16

bot190 commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -70,6 +70,19 @@ public SwerveDriveMap getSwerveDriveMap() {
+    public ShooterMap getShooterMap() {
+        final PIDSparkMax shooterMoter = new PIDSparkMax(13, MotorType.kBrushless);
+        final PIDSparkMax shooterLoadingMoter = new PIDSparkMax(14, MotorType.kBrushless);
+        final double shootWheelRadius = 2.0; // Radius in inches
```
Please use diameter not radius

bot190 commented on src/main/java/frc/robot/maps/RobotMap.java:
```diff
@@ -136,6 +136,41 @@ public BooleanSupplier getOutsideLimit() {
 
+    public class ShooterMap {
+        private SmartMotorController shootMotor;
+        private SmartMotorController loadingMotor;
```
This shouldn't be here.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,105 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
```
Please remove the WPI copyright. We're not assigning our copyright to them.

bot190 commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -70,6 +70,19 @@ public SwerveDriveMap getSwerveDriveMap() {
     }
 
+    @Override
+    public ShooterMap getShooterMap() {
```
The shooter also has a Rev throughbore encoder we can use for the shooter PID.

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  public CommandBase fire() {
+    return sequence("Shoot", setLoadingSpeed(LOADING_SPEED), new WaitCommand(SHOOT_TIME), setLoadingSpeed(0.0));
```
Could this just be a `startEnd().withTimeout(SHOOT_TIME)`?

bot190 replied:
This really shouldnt be here at all. The transporter is responsible for pushing a ball into the shooter.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,105 @@
+
+  @Override
+  public void periodic() {
+    double velocity = shootEncoder.getRate() * VEL_MUL;
```
This shouldn't be necessary. We can do the conversion in the robot map.

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,105 @@
+
+  public CommandBase setTargetHub(HubSpeed hub) {
+    return instant("Set Default Speed", () -> {
+      shooterMotor.setSetpoint(hub.getSpeed() * MAX_RPM);
```
Why not put the actual speeds in the enum directly?

bot190 commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,105 @@
+    return sequence("Shoot", setLoadingSpeed(LOADING_SPEED), new WaitCommand(SHOOT_TIME), setLoadingSpeed(0.0));
+  }
+
+  public CommandBase setLoadingSpeed(double speed) {
+    return instant("Set Loading Speed", () -> {
+      loaderMotor.set(speed);
+    });
```
The shooter is not responsible for loading a ball into the shooter at all. This command should go away.

lilbites21 committed:
Delete extra file, answered question comment

lilbites21 committed:
Cleaned up fire

lilbites21 committed:
Not my job, got rid of it.

lilbites21 committed:
Get double supplier now in command

lilbites21 committed:
Moved constant math into enumerator

lilbites21 committed:
Deleted all instances of velMul and loading motor

lilbites21 committed:
Added the second shoot motor

lilbites21 committed:
Changed radius to diameter

lilbites21 committed:
Deleted Unnecessary comment

jeffaburke committed:
Merge branch 'main' into Shooter

jeffaburke committed:
Added velocity into robot map

msoucy commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -69,6 +71,20 @@ public SwerveDriveMap getSwerveDriveMap() {
 
+    @Override
+    public ShooterMap getShooterMap() {
+        final PIDSparkMax shooterMoter1 = new PIDSparkMax(15, MotorType.kBrushless);
```
"Motor" misspelled

msoucy commented on src/main/java/frc/robot/maps/GastonMap.java:
```diff
@@ -69,6 +71,20 @@ public SwerveDriveMap getSwerveDriveMap() {
+        final double shootWheelDiameter = 2.0; // Diameter in inches
+
+        // Convert rpm to ft/s
+        final double velocityMultiplier = shootWheelDiameter * Math.PI / (60 * 12);
```
Use a line like `shooterMotor1.getEncoder().setPositionConversionFactor(x)` (and velocity conversion factor) for this instead of passing through the map

jeffaburke replied:
Not needed anymore

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,81 @@
+  public Shooter(ShooterMap shooterMap) {
+    shooterMotor = shooterMap.getShooterMotor();
+    shootEncoder = shooterMotor.getEncoder();
+    velocity = shooterMap.getVelocity();
```
Is this actually a velocity? It can't ever change

jeffaburke replied:
Not needed anymore

msoucy commented on src/main/java/frc/robot/subsystems/Shooter.java:
```diff
@@ -0,0 +1,81 @@
+    private double speed;
+
+    private HubSpeed(double speed) {
+      this.speed = speed * Shooter.getMaxRpm();
```
I may be wrong, but I think that you can use `MAX_RPM` here directly.

lilbites21 replied:
I think you would be correct, if its not already fixed I will.

jeffaburke committed:
Velocity got purged

jeffaburke committed:
Stop shooter

JoeJanickiChSh committed:
Make motor2 follow motor1

JoeJanickiChSh committed:
Merge main into Shooter

JoeJanickiChSh committed:
Remove unused imports

JoeJanickiChSh committed:
Remove smurf names

JoeJanickiChSh committed:
Add external encoder

JoeJanickiChSh committed:
Remove finishedWhen and use until

JoeJanickiChSh committed:
Remove copyright notice

JoeJanickiChSh committed:
Default to lower hub

JoeJanickiChSh committed:
Implement software PID

From SysID test

JoeJanickiChSh committed:
Add feedforward to PID controller

JoeJanickiChSh committed:
Test SystemID

JoeJanickiChSh committed:
Add feedforward calculation

JoeJanickiChSh committed:
Use a PID controller in the subsystem

JoeJanickiChSh committed:
Use correct units for PID

JoeJanickiChSh committed:
Set encoder distance per pulse

JoeJanickiChSh committed:
Use 1x encoding

JoeJanickiChSh committed:
Get PID to work

JoeJanickiChSh committed:
Get speed for low hub

msoucy committed:
Rename Valkyrie Map to proper name

JoeJanickiChSh committed:
Merge main into shooter

JoeJanickiChSh committed:
Add shooter functionality back into shooter

This was painful

JoeJanickiChSh committed:
Fix tab size

JoeJanickiChSh committed:
Make a shoot command

JoeJanickiChSh committed:
Uncomment drive command

JoeJanickiChSh committed:
Tabs

JoeJanickiChSh committed:
Put PID controller into robot map

JoeJanickiChSh committed:
Change name of setTargetHub
